### PR TITLE
Introduce scope limiter on cascading rules to enforce scope

### LINKED ISSUE #761: Introduce reverse matches on cascading rules to enforce scope
### Is your feature request related to a problem?
As a pentester, I want to gain confidence in letting run secureCodeBox cascading rules unattended, so that I have assurance that secureCodeBox will only scan in-scope targets given the engagement.

Currently, cascading rules are triggered based on whether a static value exists in the finding.
```yaml
spec:
  matches:
    anyOf:
      - category: "Subdomain"
        osi_layer: "NETWORK"
  scanSpec:
    scanType: "nmap"
    parameters:
      # Target Port of the finding
      - "{{location}}"
```

It is however not possible to verify whether the matched subdomain is present in the given scope.

A practical example engagement scope:
  - `example.com`
  - `subdomain.example.com`
  - `12.34.56.78/24`

A pentester would like to investigate whether more subdomains exist so runs AMASS, it returns `example.com`, `subdomain.example.com`, and `subdomain2.example.com`. Nmap is triggered on all of them, while `subdomain2.example.com` was out-of-scope. Furthermore, it should have been verified whether `example.com` and `subdomain.example.com` even resolve to an IP in `12.34.56.78/12`.

### Describe the solution you'd like
In the cascading rule spec, we can introduce a field allowing rule developers to do reverse matching on scan labels.

```yaml
metadata:
  labels:
    "engagement.scope/domains": "example.com,subdomain.example.com"
    "engagement.scope/cidr": "12.34.56.78/24"
```

Then in cascading rule yaml:
```yaml
spec:
  matches:
    scanSelector:
      allOf:
        - key: "engagement.scope/domains"
          operator: Contains
          value: {{attributes.host}}
        - key: "engagement.scope/cidr"
          operator: InCIDR
          value: {{attributes.ip}}
    anyOf:
      - category: "Subdomain"
        osi_layer: "NETWORK"
  scanSpec:
    scanType: "nmap"
    parameters:
      # Target Port of the finding
      - "{{location}}"
```

`scanSelector`  would be a similar field to the Kubernetes-native `labelSelector`, but fully interpreted in the cascading scans hook. This also allows us to define special operators (like Contains or InCIDR).

`allOf` is used in the example above. Would `anyOf` or `noneOf` have any further use cases?

In any case, I'd like some feedback on the problem and potential solution :thinking: . What do you think?

### PR DESCRIPTION
**Pull request by EndPositive**
Closes #761

## Description
This PR introduces `ScopeLimiter` on `CascadeSpec` so to match expressions can be added to enforce scope on an engagement. In #761, this field was thought to be added to the `CascadingRuleSpec`, but I've decided against it because the scope is usually defined on an engagement, instead of on a specific rule. This results in the following scan spec:

```yaml
apiVersion: "execution.securecodebox.io/v1"
kind: Scan
metadata:
  name: "nmap-scan"
  annotations:
    scope.cascading.securecodebox.io/domain: "secura.com" 
spec:
  cascades:
    scopeLimiter:
      allOf:
      - key: "scope.cascading.securecodebox.io/domains"
        operator: "In"
        values: ["{{attributes.hostname}}"]
  scanType: "nmap"
  parameters:
    - "-p-"
    - "136.144.234.6"
```

This scan results in a finding with the hostname `136-144-234-6.colo.transip.net`, which I don't want to scan, because my scope is only the domain `secura.com`. With the PR, the `scopeLimiter` field is evaluated and cascading scans would not be triggered.

To ensure that you can define the selector independent of the finding format, I've added a new field `scopeLimiterAliases` in the `ParseDefinitionSpec`. This field allows you to define aliases per scanner. This field should be addable through `values.yaml` in each scanner (not yet implemented). The following alias could be added to nmap's parse definition.

```yaml
scopeLimiterAliases:
  hostname: "{{attributes.hostname}}"
```

The scan definition would be:
```yaml
scopeLimiter:
  allOf:
  - key: "scope.cascading.securecodebox.io/domains"
    operator: "In"
    values: ["{{$.hostname}}"]
```

Since a hostname might not be available for every parser, I've added the field `ValidOnMissingRender` so that an expression containing a mustache value which could not be rendered, resolves to `true`. This behavior is disabled by default. The following expression would result to `true`.

```yaml
scopeLimiter:
  validOnMissingRender:  true
  allOf:
  - key: "scope.cascading.securecodebox.io/domains"
    operator: "In"
    values: ["{{invalid_field}}"]
```

The `allOf`, `anyOf`, and `noneOf` are **independently** supported. The following operator are supported: `In`, `NotIn`, `Exists`, `DoesNotExist`, `Contains`, `DoesNotContain`, `InCIDR`, `NotInCIDR`, `SubdomainOf`, `NotSubdomainOf`.

I'd like some feedback on this particular approach :smile:.

### Todo

* [x] Initial concept implementation
* [x] Create documentation PR
* [x] Rename field `scanAnnotationSelector`
* [x] Add values field for `selectorAttributeMappings`.

### Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [x] Make sure `npm test` runs for the whole project.
* [ ] Make codeclimate checks happy

### CODE REVIEW

**Comment by malexmave** [comment]
Thanks for your effort already! Some initial thoughts on what you wrote:

I like the idea of shifting this from the CascadingScan definition to the Scan definition, makes a lot more sense there in principle, and it should work fine for single-stage cascading scans (i.e., a scan triggers a CascadingScan, and then you're done). In this case, a single generic cascading scan definition is sufficient for multiple engagements.

As soon as you have cascading scans that trigger additional cascading scans, it gets a bit more painful if you want to keep enforcing scope further down the chain. You'll need to write a custom cascading scan for each engagement after all, to specify the scope for further cascades in the ScanSpec, and you'll also need to start working with the hook selector features you recently contributed to avoid having cascading rules from different engagements trigger on the same scan, so at that point we're back to where we were before moving this to the Scan definition in terms of the complexity of setting things up. But since it is never worse than previously specified, and sometimes better, *and* it makes more intuitive sense, I think that moving it is a good thing.

One place where this new behavior may be worse than specifying it on a per-hook basis is this: If I have some complex engagement rules that could only be modelled as a nested boolean formula ("(A AND B) OR (C AND D)"), this presumably cannot be represented with the syntax you are building (or can you have an `anyOf` with `allOf` as a child?). With the hook-based notation, you could just have created two cascading rules to represent this, which would have only required a single run of the scanner. With the scan-based notation, you would have to run the scanner two separate times with different scope definitions. However, I am simply saying this here so that everyone is aware of this limitation - I still think that the scan-based notation is probably superior, since such complex engagement rules are probably uncommon.

How will this interact with inheritance on cascading scans? Standard inheritance behavior (turned off by default), or will this require anything more fancy?

The new logic will be in addition to the old one (invasive/non-invasive etc.) and constraints from both will be combined, correct?

> Since a hostname might not be available for every parser, I've added the field ValidOnMissingRender so that an expression containing a mustache value which could not be rendered

Since the definition for what you want to match is specified on a single scan, which has a single parser (which presumably will either always or never have a specific field in the results), how can a case appear where the field is not defined? If the field is set to `undefined` / `null` / the empty string, that is one thing (which is why we have the `$.HostOrIP` helper), but shouldn't at least the rendering of mustache be either always or never successful?

> The `allOf`, `anyOf`, and `noneOf` are independently supported.

What do you mean with "independently supported"?

> The following operator are supported: In, NotIn, Exists, DoesNotExist, Contains, DoesNotContain, InCIDR, NotInCIDR.

What happened to the domain-based operator (`subdomainOf`)? Modeling this as "Contains" / "DoesNotContain" can be misleading, since you may have external service providers that include the domain as part of their domain name (`securecodebox.io.services.azure.com` or something like that). In that case, having an operator that is aware of the semantics of top/2nd level domains may be useful.

**Comment by EndPositive** [comment]
Hello again :smile:,

For Secura, it makes sense to enforce scope throughout an entire engagement and thus force inherit the scope selector and not allow overriding it in cascading rules (i.e. no merge/perge issues). I think this tackles most of the points you discussed about moving the selector from hook to scan.

Some comments below:

>  cascading rules from different engagements trigger on the same scan,
 
I'm not really sure what you mean here. We have a default set of cascading rules and use labels to match rules per customer or engagement.

> The new logic will be in addition to the old one (invasive/non-invasive etc.) and constraints from both will be combined, correct?

Definitely. The labels are static matches using Kubernetes-native label selectors. These are evaluated while retrieving the rules. The annotation selectors introduced in this PR are evaluated in "runtime" and match on the scan's annotations instead of on just the findings. 

> Since the definition for what you want to match is specified on a single scan, which has a single parser (which presumably will either always or never have a specific field in the results), how can a case appear where the field is not defined?

The idea for `SelectorAttributeMappings` is that you can use the same `ScanAnnotationSelector` template values for multiple different scanners. An example is that you might want to select based on IP, which Nmap gives in `attributes.ip` and sslyze in `attributes.ip_address`. The `$.HostOrIP` variables are far too generic and oblivious of the scantype/scanner. Furthermore, these are hardcoded.. The goal with the mappings is that it should always be rendered. If it cannot be rendered, the behaviour is defined by `ValidOnMissingRender`.

> or can you have an anyOf with allOf as a child?

Nope, not sure if that would be possible to express in CRD.

> What do you mean with "independently supported"?

Code is currently like below. To support more complex scenario's we could allow specifying `allOf`, `anyOf`, and `noneOf` in combination and evaluating them together somehow (maybe simple AND)?

```ts
if (scanAnnotationSelector.allOf !== undefined) return scanAnnotationSelector.allOf.every(validateRequirement);

if (scanAnnotationSelector.anyOf !== undefined) return scanAnnotationSelector.anyOf.some(validateRequirement);

if (scanAnnotationSelector.noneOf !== undefined) return !scanAnnotationSelector.noneOf.some(validateRequirement);
```

> What happened to the domain-based operator (subdomainOf)?

I'll implement it once you like the approach :smile:.

**Comment by malexmave** [comment]
> For Secura, it makes sense to enforce scope throughout an entire engagement and thus force inherit the scope selector and not allow overriding it in cascading rules (i.e. no merge/perge issues).

Hmm, I see. In cases where you can be sure that all scans that should trigger cascading scans will always return a hostname / IP that you can match on, this should work fine. And I guess there aren't any scanners that you would want to trigger additional network scans that will not return this. So, makes sense. However, the thing you describe (not allowing overriding) would make the inheritance behavior of the scope limiters different from that of the other inherited values... We had discussed that can of worms in #789. This may be a point where some input by @J12934 may be useful, as he is usually our "consistency evangelist" 😁. (Or am I misunderstanding you somewhere?)

>> cascading rules from different engagements trigger on the same scan,
> I'm not really sure what you mean here. We have a default set of cascading rules and use labels to match rules per customer or engagement.

I was still thinking about cascading rules that specify different scopes for their sub-scans - either without inheritance, or with further limitations being added down the line, like saying "You can scan all these IPs, but please leave the following ports alone for follow-up scans". Or you may have one scan that discovers hosts (e.g., based on DNS), followed by an nmap on in-scope systems, followed by a more detailed scanner that operates on specific ports, and only certain ports are in-scope. You cannot specify the ports-in-scope on the initial scan, since there is no port information anywhere in its output and so specifying the port requirements would break the cascade unless you play tricks with your "aliasing" + "ignore this if templating fails" logic. This may be a use case where you would need to specify new scope limiters in your cascading scans ScanSpec, which would mean that you need per-engagement custom cascading rules. 

tldr: Its another edge case that can show up, but its solvable with hook selectors and not necessary in 99% cases, so its fine the way you propose to do it. (You may notice that I think a lot about strange edge cases)

> The idea for SelectorAttributeMappings is that you can use the same ScanAnnotationSelector template values for multiple different scanners. An example is that you might want to select based on IP, which Nmap gives in attributes.ip and sslyze in attributes.ip_address.

I see, so its aliasing that is used in combination with the inherited scope limiters. Makes sense!

>>or can you have an anyOf with allOf as a child?
> Nope, not sure if that would be possible to express in CRD.

Good, its a weird edge case anyway and not necessary. 😁 

> To support more complex scenario's we could allow specifying allOf, anyOf, and noneOf in combination and evaluating them together somehow (maybe simple AND)?

~Hmm... Shouldn't any linear AND combination of allOf, anyOf and noneOf be expressible by a simple allOf with well-chosen matches inside of it?~ After thinking about it: Nope, it shouldn't. Many are, but not all of them.

Combining all of them via AND may actually also resolve another problem I have with your code right now, which is that specifying all three operators is valid (in the sense that the cluster will accept a YAML file with them and not throw errors), but the code snippet you posted will just silently ignore most of them in favor of the first type it matches (i.e., if an allOf operator is present, anyOf and (more crucially) noneOf are ignored). ANDing them together would provide us with an intuitive and easy to document solution to this.

**Comment by EndPositive** [comment]
> consistency evangelist

Hahah yeah I get you. I'm down to implement it anyways, so let me know.

> in the sense that the cluster will accept a YAML file with them and not throw errors

Yeah I was actually wondering how to specify mutual exclusiveness in CRD, but I don't think it's possible. I'll implement it with AND.

**Comment by malexmave** [comment]
Just talked to @J12934. He is fine with the proposed inheritance behavior. :)

One other point he raised: `engagement.scope/domain` as an annotation name is violating the policies of Kubernetes, which says that if you do something that looks like a domain, you must own that domain. He also raised the point that if you want scope definition to be immutable, but you are using annotations to store information about them, then these annotations by default fall under the standard annotation inheritance rules (i.e., they can be modified or deleted).

He proposes `scope.cascadingscans.securecodebox.io/*` as the namespace for scope annotations (this also makes clear that the scope is only defined for cascading scans, and does not enforce anything on the first scan, i.e., if you set a scope of example.com on an "nmap google.com" scan, it will not prevent the scan). This would also allow you to solve one problem with another and define special handling for these annotations if you really want to make sure that these annotations are inherited without modifications (if that is the behavior you want to go for).

**Comment by EndPositive** [comment]
Awesome. I'll take a look at annotation handling. Should `ScanAnnotationSelector` also work for annotations that are not in that namespace? If not, should we consider renaming this field?

**Comment by malexmave** [comment]
Hmm. An argument could be made that it should be called `scopeSelector` or `scopeLimiter` or something like that. Is there any reasonable situation in which you would want to use the ScanAnnotationSelector to select on non-scope-related annotations? I can't come up with one on short notice, but maybe you know one.

If there isn't, and if you want to force inheritance on scope-related annotations no matter if other annotations are inherited, it probably makes sense to:
1. Rename the selector to something with "scope" in the name
2. Check in the cascading hook that it only matches on annotations in the scope namespace (and throw an error and stop execution if it doesn't, because this should be considered a syntax error)
3. Always inherit all annotations in that namespace, no matter if other annotations are inherited
4. Potentially prevent new annotations from being added to that namespace in the cascading definitions (not sure about this - do you think this would be necessary / a good idea? It's the only way to enforce immutability)

**Comment by EndPositive** [comment]
@malexmave implemented your proposal. See the test `scope annotations should be completely immutable`, showcases behavior really well. Currently, any scope annotations added by a cascading rule's `scanLabels` are silently discarded. Should we throw errors for this?

Still thinking about a better name for `scanAnnotationSelector`.

**Comment by malexmave** [comment]
> Currently, any scope annotations added by a cascading rule's scanLabels are silently discarded. Should we throw errors for this?

I would argue that "failing loudly" (stopping execution and printing an error message) would be the right move here, as silently discarding information will lead to impossible-to-track-down errors for users.

**Comment by EndPositive** [comment]
@malexmave @J12934 ready for review

**Comment by malexmave** [comment]
I will review this soon. Just a quick point so that we can avoid having the headache again: Can you make sure the helm docs are up to date so that we don't have the issue again where every new pull request runs the "Update helm docs" action? 😁 

The resulting changes from the current state of the PR can be seen in https://github.com/secureCodeBox/secureCodeBox/commit/0b5c09b7eff64757077a63d9da2b55b8d9acc722 (I pushed it to a branch in our repo to run the CI).

**Comment by EndPositive** [comment]
Cherry-picked your commit. Shouldn't the helm-docs CI action run when the PR is merged? Avoids having to do it manually.

**Comment by malexmave** [comment]
The helm-docs CI will run when it is merged, but it will fall on its face because the main branch is protected and it can't commit to it 😁.

Looks like there are some helm-docs-related merge conflicts now, but those should be rather straightforward to resolve, I hope.

File: hooks/cascading-scans/hook/reverse-matches.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    throw new Error(`values may not be undefined`)
+  }
+}
+
+function operatorIn({lhs, rhs}: Operands): boolean {
+  return rhs.includes(lhs);
+}
+function operatorExists({lhs, rhs}: Operands): boolean {
+  return lhs !== undefined;
+}
```
Comment (malexmave) [inline hooks/cascading-scans/hook/reverse-matches.ts:166]: This is coming back to the point of "calling the variables `lhs` / `rhs` makes this really hard to understand", but the way I understand things, `rhs` is the list of findings that are templated based on the "values" field in the scope limiter, while `lhs` is the value from the referenced limiter annotation. So, in which case can the value from the annotation be undefined? Or rather, in which case would it be useful to test for it, since this value is fixed? Should this be `rhs !== undefined`?

File: hooks/cascading-scans/hook/reverse-matches.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+function operatorContains({lhs, rhs}: Operands): boolean {
+  const valueArray = lhs.split(",");
+  return rhs.every(value => valueArray.includes(value));
+}
```
Comment (malexmave) [inline hooks/cascading-scans/hook/reverse-matches.ts:171]: I'm still missing some documentation for what the operators are actually supposed to do, so I can't really tell if this is correct behavior. The way I understand it, this operator receives a list of findings in the `value` key (as does every other), but it interprets the provided scope from the scope annotation value as a comma-separated list of values. It then checks if every finding is represented in the comma-separated list of in-scope targets. Is this correct?

I think I'll need some documentation for the expected behavior of these operators before I can continue reviewing this - having to guess at these things is a recipe for introducing unexpected behavior. This documentation should ideally be both inline in this file, and (as I requested in the docs PR) in the written user documentation.

**Review by malexmave — CHANGES_REQUESTED** [review]
Submitting some comments from a review-in-progress. I have to pause the review now until the questions from these comments here are addressed, as I don't think that I can usefully review this any further until then.

File: hooks/cascading-scans/hook/hook.ts [hunk 3]
```diff
@@ -86,9 +106,31 @@ export function getScanChain(parentScan: Scan) {
...
+      finding,
+      parseDefinition.spec.scopeLimiterAliases,
+    );
+
+    if (!reverseMatches) {
+      console.log(`Cascading Rule ${cascadingRule.metadata.name} not triggered as scope limiter did not pass`);
+      console.log(`Scan annotations ${JSON.stringify(parentScan.metadata.annotations)}`);
+      console.log(`Scope limiter ${JSON.stringify(parentScan.spec.cascades.scopeLimiter)}`);
+      console.log(`Scope limiter aliases ${JSON.stringify(parseDefinition.spec.scopeLimiterAliases)}`);
+      console.log(`Finding ${JSON.stringify(finding)}`);
```
[the remark this answers is not in the document]
Comment (malexmave) [reply hooks/cascading-scans/hook/hook.ts:130]: Okay, great, then this is fine :)

File: hooks/cascading-scans/hook/reverse-matches.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  if (scopeLimiter === undefined) return true;
+
+  function validateRequirement({key, operator, values}: ScopeLimiterRequirement) {
+    if (!key.startsWith(`${scopeDomain}`)) {
+      throw new Error(`key '${key}' is invalid: key does not start with '${scopeDomain}'`);
+    }
+
+    const { operator: operatorFunction, validator: validatorFunction } = operatorFunctions[operator];
+    if (operatorFunction === undefined) {
+      throw new Error(`Unknown operator '${operatorFunction}'`);
```
[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts:57]:
```js
const { operator: operatorFunction, validator: validatorFunction } = operatorFunctions[operator];
```

Means unpack the result of `operatorFunctions[operator]` and map `operator` -> `operatorFunction` and `validator` -> `validatorFunction` to constants.

[the remark this answers is not in the document]
Comment (malexmave) [reply hooks/cascading-scans/hook/reverse-matches.ts:57]: Yes, that's not the issue, the issue is that if-statement:

```javascript
if (operatorFunction === undefined) {
      throw new Error(`Unknown operator '${operatorFunction}'`);
}
```
You will only ever get into the body and the throw if `operatorFunction` is `undefined`, so you will only ever print `undefined` in the Error (I think?).

[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts:57]: Ahh, yeah you're right :thinking: .

File: scanners/amass/templates/amass-parse-definition.yaml [hunk 5]
```diff
@@ -12,3 +12,5 @@ spec:
   ttlSecondsAfterFinished: {{ .Values.parser.ttlSecondsAfterFinished }}
   env: 
     {{- toYaml .Values.parser.env | nindent 4 }}
+  scopeLimiterAliases:
```
[the remark this answers is not in the document]
Comment (EndPositive) [reply scanners/amass/templates/amass-parse-definition.yaml:15]: 3a25ec18 - Add `scopeLimiterAliases` to new scanner template

My other PRs only introduced new CRD fields in scan and hook spec, so i think this template is good.

File: hooks/cascading-scans/hook/hook.ts [hunk 6]
```diff
@@ -86,9 +106,31 @@ export function getScanChain(parentScan: Scan) {
...
+function getScansMatchingRule(
+  parentScan: Scan,
+  findings: Array<Finding>,
+  cascadingRule: CascadingRule,
+  parseDefinition: ParseDefinition,
+) {
   const cascadingScans: Array<Scan> = [];
   for (const finding of findings) {
+    // Check if the scan matches for the current finding
+    const reverseMatches = isReverseMatch(
```
[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/hook.ts]: 88373f18 Rename `reverse-matches.ts` to `scope-limiter.ts` and `isReverseMatch` to `isInScope`

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: 548ba43b - Error operator name instead of operator function

File: hooks/cascading-scans/hook/reverse-matches.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    scopeLimiter.allOf !== undefined ? scopeLimiter.allOf.every(validateRequirement) : true,
+    scopeLimiter.anyOf !== undefined ? scopeLimiter.anyOf.some(validateRequirement) : true,
+    scopeLimiter.noneOf !== undefined ? !scopeLimiter.noneOf.some(validateRequirement) : true,
+  ].every(entry => entry === true);
+}
+
+interface Operands {
+  lhs: string,
+  rhs: string[],
+}
```
[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: cccd6220 - Rename lhs and rhs to their semantic names

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 1 above)
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: It is indeed testing whether the annotation is undefined. In other words, whether the specified annotation exists on the scan.

With some further thoughts, I can't think of a scenario where you would want to use this. You define them both before you start the scan..

I made this because the operator also exists in default Kubernetes label selector. We can delete it if you can't think of a good use-case.

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 2 above)
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: I've updated my docs PR with short explanation for each operator. I'll copy them to source code as soon as you approve them in the docs PR, to prevent double work.

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 1 above)
Comment (malexmave) [reply hooks/cascading-scans/hook/reverse-matches.ts]: Yeah, I don't really know any good use case for that as long as adding additional scope limiters during the cascade is not possible. We should probably remove it, here and in the documentation PR.

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 2 above)
Comment (malexmave) [reply hooks/cascading-scans/hook/reverse-matches.ts]: Great, thank you. The explanations in the docs PR are fine, feel free to copy-paste them here.

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 1 above)
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: 3f1dbd5f - Get rid of redundant operators

File: hooks/cascading-scans/hook/reverse-matches.ts (hunk 2 above)
Comment (EndPositive) [reply hooks/cascading-scans/hook/reverse-matches.ts]: 0c1d8c9 - Add inline documentation for every operator

**Review by malexmave — APPROVED** [review]
After a fruitful and long set of discussions, debugging, testing and changes, this PR is finally ready to merge 🎉. Thanks again @EndPositive for the great work, long discussions and many changes you implemented.

**Review by malexmave — APPROVED** [review]

**Comment by J12934** [comment]
Updated the branch & then resolved the merge issues in the unit tests. Seems to be working now.
Will merge as soon as the other checks are passing and completed.

We should really tackle the code complexity in the cascading hook soon. The test file is already at 2800 lines of code 😬
Does anybody have a good overview over the current hooks code and ideas on how it could be refactored to be simpler?

**Review by J12934 — APPROVED** [review]

**Comment by EndPositive** [comment]
@J12934, I think it would be a good idea to refactor the test file to get rid of the snapshot testing. Also, we should probably group the hook tests in the same manner as the scope limiter tests.

Regarding the complexity of the hook itself. Maybe that should be combined with #789..
